drivers : espi : saf Add Microchip eSPI SAF driver

scottwcpg opened the pull request:
Add Microchip eSPI SAF driver supplementing MCHP eSPI
driver for MEC15xx SoC family.

Signed-off-by: Scott Worley <scott.worley@microchip.com>

albertofloyd commented on drivers/espi/Kconfig.xec:
```diff
@@ -37,4 +37,19 @@ config ESPI_PERIPHERAL_UART_SOC_MAPPING
+config ESPI_SAF
+	bool "XEC Microchip ESPI SAF driver"
+	depends on ESPI_FLASH_CHANNEL
+	default y
```
Do not enable by default, some board designs do not support SAF

scottwcpg replied:
Change to "n" in next rebase

franciscomunoz commented on drivers/espi/espi_saf_mchp_xec.c:
```diff
@@ -0,0 +1,898 @@
+		fcfg++;
+		totalsz += fcfg->flashsz;
+	}
+	/* Program CS1 configuration (same as CS0 if only one device) */
```
Add a space after }

scottwcpg replied:
Done in next rebase

andrewboie commented on drivers/espi/CMakeLists.txt:
```diff
@@ -3,4 +3,10 @@
-zephyr_library_sources_ifdef(CONFIG_USERSPACE           espi_handlers.c)
+zephyr_library_sources_ifdef(CONFIG_USERSPACE		espi_handlers.c)
+
+if(CONFIG_ESPI_SAF AND CONFIG_ESPI_XEC AND CONFIG_ESPI_FLASH_CHANNEL)
```
isn't `zephyr_library_sources_ifdef(CONFIG_ESPI_SAF` sufficient here?

scottwcpg replied:
Done

andrewboie commented on include/drivers/espi_saf.h:
```diff
@@ -0,0 +1,522 @@
+ * @retval -EBUSY eSPI flash channel is not ready or disabled by master.
+ * @retval -EIO General input / output error, failed request to master.
+ */
+__syscall int espi_saf_flash_erase(struct device *dev,
```
where are the verification functions for this and the other system calls that are added in this PR?

scottwcpg replied:
Is it acceptable to add verification tests in this PR?

andrewboie replied:
from my perspective it's mandatory.

andrewboie replied:
either that, or drop the `__syscall`.

scottwcpg replied:
Ok. Making sure the test code is still working after pulling the latest samples/drivers/espi changes.

scottwcpg replied:
Pushed rebase containing SAF tests added to samples/drivers/espi. This rebase defaults samples/drivers/espi to its default settings (no SAF). Edit the *.conf files and enable QMSPI and SAF to build the application with SAF support.

tyrantlizard77 reviewed (changes requested):
In addition to issues noted inline, this PR introduces many new APIs without providing tests.

This PR appears to introduce a new driver subsystem, not just a new driver.

dcpleung commented on drivers/espi/Kconfig.xec:
```diff
@@ -37,4 +37,19 @@ config ESPI_PERIPHERAL_UART_SOC_MAPPING
 
+config ESPI_SAF
+	bool "XEC Microchip ESPI SAF driver"
+	depends on ESPI_FLASH_CHANNEL
```
Does this also need to depend on `ESPI_XEC`?

andrewboie replied:
it's inside a `if ESPI_XEC` block.

nashif commented on samples/drivers/espi/prj.conf:
```diff
@@ -1,4 +1,7 @@
 CONFIG_PRINTK=y
-
+# Uncomment to test SAF API's
+#CONFIG_ESPI_SAF=y
```
please do not add commented options, create a new configuration with SAF enabled and configure it into sample.yaml so that it is built by CI.

albertofloyd replied:
@scottwcpg Enable it only for EVB

nashif commented on samples/drivers/espi/prj_mec1501modular_assy6885.conf:
```diff
@@ -7,3 +7,7 @@ CONFIG_LOG_PROCESS_THREAD_SLEEP_MS=100
 # Sample code doesn't handle ACPI host communication
 CONFIG_ESPI_PERIPHERAL_HOST_IO=n
+# Uncomment to test SAF EC portal API's
+#CONFIG_ESPI_SAF=y
```
ditto

scottwcpg replied:
@albertofloyd recommended SAF tests be enabled only for the evaluation board (mec15xxevb_assy6853).
Removed commented out SAF config items from the prj.conf and prj_mec1501modular_assy6885.conf.
Uncommented SAF config enables in prj_mec15xxevb_assy6853.conf
Updated the SAF driver Kconfig to be disabled by default.

nashif commented on samples/drivers/espi/prj_mec15xxevb_assy6853.conf:
```diff
@@ -0,0 +1,13 @@
+# eSPI + mec15xxevb_assy6853
```
what is the point of this file? we already have a board specific .conf file already.

albertofloyd replied:
@nashif 
prj.conf is for any non-mchp board that may support eSPI in the future. 
mec15xxevb_assy6853 configuration is main validation vehicle board.
mec15modular is very limited in standalone mode for validation, requires to be used in conjunction with Intel RVP
We need separate files to enable different features in each board

nashif commented on dts/arm/microchip/mec1501hsz.dtsi:
```diff
@@ -219,6 +219,17 @@
+				0x40070000 0x400
+				0x40071000 0x400>;
+			label = "ESPI_SAF_0";
+			#address-cells = <1>;
```
why are those commented out?

mbolivar-nordic replied:
> why are those commented out?

Those aren't comments. It's a property named `#address-cells`.

nashif reviewed (changes requested):
This PR needs a rebase and needs a test/sample that builds in CI, currently it does not build, most likely due to changes to the device structure (constification).

Also, this PR is 1 commit that adds a subsystem, a driver, changes to SoC and then changes to a sample, it would be great if you could split the commits.

scottwcpg committed:
drivers: espi: Microchip eSPI add SAF support

Add ESPI SAF features to the Microchip eSPI driver as
a separate file. ESPI SAF depends upon the core eSPI
driver adding the ability to attach the system SPI
flash to the EC eSPI endpoint instead of the host
eSPI controller.

Signed-off-by: Scott Worley <scott.worley@microchip.com>

scottwcpg committed:
samples: drivers: espi: Add simple SAF tests

Update sample espi driver test to exercise some SAF functionality
based on the limited hardware available. Note the SAF test will
erase a section of the flash. SAF tests are enabled for the EVB only.

Signed-off-by: Scott Worley <scott.worley@microchip.com>

mbolivar-nordic commented on dts/bindings/espi/microchip,xec-espi-saf.yaml:
```diff
@@ -0,0 +1,48 @@
+# Copyright (c) 2019 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+description: Microchip ESPI SAF controller
+
+# Copyright (c) 2020 Microchip Technology Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+description: Microchip ESPI SAF controller
```
Some copy/paste errors happened here.

mbolivar-nordic replied:
Sorry I was not precise enough. The description has been fixed, but the SPDX line is still duplicated and the copyrights are not all together.

The top of this file should look like this:

```yaml
# Copyright (c) 2019 Intel Corporation
# Copyright (c) 2020 Microchip Technology Inc.
# SPDX-License-Identifier: Apache-2.0

description: Microchip ESPI SAF controller

compatible: "microchip,xec-espi-saf"
```

scottwcpg replied:
Hi, fixed it.

mbolivar-nordic commented on dts/bindings/espi/microchip,xec-espi-saf.yaml:
```diff
@@ -0,0 +1,48 @@
+
+    consec_rd_timeout:
+      type: int
+      description: timeout after last read to resume supended operations
```
units?

mbolivar-nordic commented on drivers/espi/espi_saf_mchp_xec.c:
```diff
@@ -0,0 +1,906 @@
+	.saf_base_addr = DT_INST_REG_ADDR_BY_IDX(0, 0),
+	.qmspi_base_addr = DT_INST_REG_ADDR_BY_IDX(0, 1),
+	.saf_comm_base_addr = DT_INST_REG_ADDR_BY_IDX(0, 2),
+#if DT_INST_NODE_HAS_PROP(0, poll_timeout)
+	.poll_timeout = DT_INST_PROP(inst, poll_timeout),
+#else
+	.poll_timeout = MCHP_SAF_FLASH_POLL_TIMEOUT,
+#endif
```
```suggestion
	.poll_timeout = DT_INST_PROP_OR(inst, poll_timeout, MCHP_SAF_FLASH_POLL_TIMEOUT),
```

mbolivar-nordic commented on drivers/espi/espi_saf_mchp_xec.c:
```diff
@@ -0,0 +1,906 @@
+#else
+	.poll_timeout = MCHP_SAF_FLASH_POLL_TIMEOUT,
+#endif
+#if DT_INST_NODE_HAS_PROP(0, consec_rd_timeout)
+	.consec_rd_timeout = DT_INST_PROP(inst, consec_rd_timeout),
+#else
+	.consec_rd_timeout = MCHP_SAF_FLASH_CONSEC_READ_TIMEOUT,
+#endif
+#if DT_INST_NODE_HAS_PROP(0, sus_chk_delay)
+	.sus_chk_delay = DT_INST_PROP(inst, sus_chk_delay),
+#else
+	.sus_chk_delay = MCHP_SAF_FLASH_SUS_CHK_DELAY,
+#endif
+#if DT_INST_NODE_HAS_PROP(0, sus_rsm_interval)
+	.sus_rsm_interval = DT_INST_PROP(inst, sus_rsm_interval),
+#else
+	.sus_rsm_interval = MCHP_SAF_FLASH_SUS_RSM_INTERVAL,
+#endif
+#if DT_INST_NODE_HAS_PROP(0, poll_interval)
+	.poll_interval = DT_INST_PROP(inst, poll_interval),
+#else
+	.poll_interval = MCHP_SAF_FLASH_POLL_INTERVAL,
+#endif
```
These can all use DT_INST_PROP_OR also, as shown in the previous comment.

mbolivar-nordic commented on drivers/espi/espi_saf_mchp_xec.c:
```diff
@@ -0,0 +1,906 @@
+#endif
+};
+
+DEVICE_AND_API_INIT(espi_saf_xec_0, DT_INST_LABEL(0),
```
Use DEVICE_DT_INST_DEFINE instead.

scottwcpg replied:
Hi. Thanks for the review.
I made the above changes, rebased to current master, and pushed it.

mbolivar-nordic reviewed (changes requested):
A few nits related to devicetree.

scottwcpg commented:
Rebased to resolve conflict in drivers/espi/CMakeLists.txt

scottwcpg commented on drivers/espi/CMakeLists.txt:
```diff
@@ -3,4 +3,10 @@
-zephyr_library_sources_ifdef(CONFIG_USERSPACE           espi_handlers.c)
+zephyr_library_sources_ifdef(CONFIG_USERSPACE		espi_handlers.c)
+
+if(CONFIG_ESPI_SAF)
```
Converted all lines to same spacing style (hard tabs).
